include/boost/math/special_functions/prime_sieve.hpp

```diff
@@ -0,0 +1,425 @@
+
+    Integer primes_in_range {static_cast<Integer>(current_upper_bound / std::log(static_cast<double>(current_upper_bound)) -
+                             current_lower_bound / std::log(static_cast<double>(current_lower_bound)))};
+
+    for(size_t i {}; i < ranges; ++i)
+    {
+        prime_vectors[i].reserve(primes_in_range);
+
+        future_manager.emplace_back(std::async(std::launch::async, [current_lower_bound, current_upper_bound, &primes, &prime_vectors, i]{
+            boost::math::detail::mask_sieve(current_lower_bound, current_upper_bound, primes, prime_vectors[i]);
+        }));
+
+        current_lower_bound = current_upper_bound + 1;
+        current_upper_bound += interval;
+    }
+
+    prime_vectors[ranges].reserve(primes_in_range);
+    future_manager.emplace_back(std::async(std::launch::async, [current_lower_bound, upper_bound, &primes, &prime_vectors]{
+        boost::math::detail::mask_sieve(current_lower_bound, upper_bound, primes, prime_vectors.back());
+    }));
+
+    for(auto &&future : future_manager)
+    {
+        if(future.valid())
+        {
+            future.get();
+        }
+    }
+
+    for(auto &v : prime_vectors)
+    {
+        resultant_primes.insert(resultant_primes.end(), v.begin(), v.end());
+    }
+}
+
+template<class Integer, class Container>
+void segmented_sieve(Integer lower_bound, Integer upper_bound, Container &resultant_primes)
+{
+    Integer limit{static_cast<Integer>(std::floor(std::sqrt(static_cast<double>(upper_bound)))) + 1};
+    std::vector<Integer> primes {};
+    primes.reserve(limit / std::log(limit));
+
+    // Prepare for max value so you do not have to calculate this again
+    if(limit < 4096)
+    {
+        boost::math::detail::linear_sieve(limit, primes);
+    }
+
+    else
+    {
+        boost::math::detail::mask_sieve(static_cast<Integer>(2), limit, primes);
+    }
+
+    boost::math::detail::segmented_sieve(lower_bound, upper_bound, primes, resultant_primes);
+}
+
+template<typename T>
+struct IsVector
+{
+    using type = T;
+    constexpr static bool value = false;
+};
+
+template<typename T>
+struct IsVector<std::vector<T>>
+{
+    using type = std::vector<T>;
+    constexpr static bool value = true;
+};
+
+template<typename T>
+constexpr bool is_vector_v = IsVector<T>::value;
+} // End namespace detail
+
+template<class ExecutionPolicy, class Integer, class Container>
+void prime_sieve(ExecutionPolicy&& policy, Integer upper_bound, Container &primes)
+{
+    static_assert(std::is_integral<Integer>::value, "No primes for floating point types");
+    BOOST_ASSERT_MSG(upper_bound + 1 < std::numeric_limits<Integer>::max(), "Type Overflow");
+
+    if(upper_bound == 2)
+    {
+        return;
+    }
+
+    --upper_bound; // Not inclusive, but several methods in boost::math::detail need to be
+
+    if constexpr (detail::is_vector_v<decltype(primes)>)
+    {
+        primes.reserve(upper_bound / std::log(static_cast<double>(upper_bound)));
+    }
```

Comment:
This really violates the philosophy of generic programming. These algorithms should not know about the container that the result is going into. The output should be an output iterator.
Otherwise we will end up with algorithms that are not generic and someone will have to fix them later.
It's also just making your work harder and the algorithm slower by managing the client's memory.
I'm not sure who I should appeal to here, but please, this is not generic programming.